lib/timers.js

```diff
@@ -1,52 +1,110 @@
 'use strict';
 
-const Timer = process.binding('timer_wrap').Timer;
+const TimerWrap = process.binding('timer_wrap').Timer;
 const L = require('internal/linkedlist');
-const assert = require('assert').ok;
+const assert = require('assert');
 const util = require('util');
 const debug = util.debuglog('timer');
-const kOnTimeout = Timer.kOnTimeout | 0;
+const kOnTimeout = TimerWrap.kOnTimeout | 0;
 
 // Timeout values > TIMEOUT_MAX are set to 1.
 const TIMEOUT_MAX = 2147483647; // 2^31-1
 
-// IDLE TIMEOUTS
+
+// HOW and WHY the timers implementation works the way it does.
+//
+// Since many parts of Node.js and user applications rely on timeouts, there may
+// be a very large amount of timeouts scheduled at any given time.
+// Therefore, it is very important that the timers implementation is very fast
+// and efficient.
+//
+// Note: It is suggested you first read though the lib/internal/linkedlist.js
+// linked list implementation. How it works is somewhat counter-intuitive to
+// most JavaScript code, and timers depend on it extensively.
 //
-// Because often many sockets will have the same idle timeout we will not
-// use one timeout watcher per item. It is too much overhead.  Instead
-// we'll use a single watcher for all sockets with the same timeout value
-// and a linked list. This technique is described in the libev manual:
-// http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#Be_smart_about_timeouts
-
-// Object containing all lists, timers
-// key = time in milliseconds
-// value = list
-var lists = {};
-
-
-// call this whenever the item is active (not idle)
-// it will reset its timeout.
-// the main function - creates lists on demand and the watchers associated
-// with them.
+// In order to be as efficient as possible, the implementation does as little as
+// possible, and only when required.
+//
+// Object maps are kept which contain linked lists keyed by their duration in
+// milliseconds.
+// The linked lists within also have some meta-properties, one of which is a
+// TimerWrap C++ handle, which makes the call after the duration to process the
+// list it is attached to.
+//
+// With this, virtually constant-time insertion (append), removal, and timeout
+// is possible in the JavaScript layer. Any one list of timers is able to be
+// sorted by just appending to it because all timers within share the same
+// duration. Therefore, any timer added later will always have been scheduled to
+// timeout later, thus only needing to be appended. Any less-than constant
+// overhead is contained within the TimerWrap's inherently faster libuv binding.
```

Comment:
Did some research, libuv uses a heap implementation, i.e. `O(log n)`. I can't comment on it too much since C is *very* foreign to me, but having a a structure like we have seems ideal.

Our implementation works out to being like a timer wheel, but it ends up being more efficient having a heap underneath than an actual timer wheel. (i.e. does not use an absurd amount of memory.) It is quite similar to stacked timer wheels.

Will update.